### PR TITLE
[ORC][Spark] - Support selected vector with ORC reader on the row and batch reader

### PR DESCRIPTION
**Pull request by pavibhai**
## What?
Currently Iceberg does not support the use of the selected vector when reading ORC Files. This requires reads on ORC to be run in compatibility mode by not setting `orc.filter.use.selected` in the presence of filter processing that is triggered via `orc.sarg.to.filter`.

Filter processing was introduced as part of [ORC-744](https://issues.apache.org/jira/browse/ORC-744) where ORC has the ability to filter out records and indicate this status in partially filtered batches using the selected vector in VectorizedRowBatch.

This PR uses the selected vector to determine valid rows when applicable.

## Why?
ORC can only operate in compatibility mode by not setting `orc.filter.use.selected`. Enabling this will further hasten the processing of rows by ignoring rows that are already filtered out in the batch.

## Tested?
New Unit tests have been added to verify the behavior.

### CODE REVIEW

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestVectorizedOrcDataReader.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            .filter(Expressions.equal("id", 3L))
+            .config(OrcConf.ALLOW_SARG_TO_FILTER.getAttribute(), String.valueOf(true))
+            .config(OrcConf.READER_USE_SELECTED.getAttribute(), String.valueOf(true))
+            .build()) {
+      Iterator<InternalRow> rows = batchesToRows(reader.iterator());
+      Assertions.assertTrue(rows.hasNext());
+      InternalRow row = rows.next();
+      Assertions.assertEquals(3L, row.getLong(0));
+      Assertions.assertEquals("c", row.getString(1));
+      Assertions.assertArrayEquals(new int[] {3, 4, 5}, row.getArray(3).toIntArray());
```
Comment (dpaani) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestVectorizedOrcDataReader.java:162]: hi @pavibhai there was a recommendation to use assertj for assertions - https://github.com/apache/iceberg/blob/master/CONTRIBUTING.md#assertj

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java [hunk 2]
```diff
@@ -209,7 +229,8 @@ private int numNullsHelper() {
     }
 
     protected int getRowIndex(int rowId) {
-      return vector.isRepeating ? 0 : rowId;
+      int row = isSelectedInUse ? selected[rowId] : rowId;
```
Comment (dpaani) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:232]: will there be a possibility of isSelectedInUse=true and selected is empty?

Comment (pavibhai) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/VectorizedSparkOrcReaders.java:232]: No, this is not a possibility. Similar to other vectors the selected vector is initialized at the start to max batch size and all of the indices in this are excepted to be within the batch size.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestVectorizedOrcDataReader.java (hunk 1 above)
Comment (pavibhai) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestVectorizedOrcDataReader.java]: done

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @pavibhai!

**Pull request merged**